### ISSUE TITLE
No way to detect connection state (or connection failure)

### ISSUE THREAD

**Issue by Stretchdude** [edrosten/libblepp#23]
is there a way, that I am missing, to prevent my Programm to call
 BLEGATTStateMachine::read_and_process_next
if the previous call to void BLEGATTStateMachine::connect was failing? (e.g. because device is gone out of range)

I loop over read_and_process_next and stderr is flooded with "Trying to read_and_process_next while disconnected".
I haven't found something like "getState" or "isConnected" or anything and BLEGATTStateMachine::state is private.
Named methods are also void, so I can not catch the error message in code.

Do you want me to write a getter for BLEGATTStateMachine::state or something else?

Regards and thanks 
Ben

**Comment by Stretchdude** [comment]
I gave BLEGATTStateMachine::read_and_process_next a boolean return value in my fork. Which returns false if connection is lost. I don't know if this is against any design idea or s.th. but it works fine for me.
Are you interested in a pull request? I'd be happy to receive a "cleaner" solution if there is anything in your mind :-)
regards
Ben

**Comment by denysvitali** [comment]
Ping @edrosten

**Comment by edrosten** [comment]
I assumed it was no longer a problem since the OP closed the issue. I'll take a look.

**Comment by edrosten** [comment]
By the looks of the code, you should either get an exception (if there was some sort of error) or a disconnected() callback if the connection was never made.

Are you in blocking or nonblocking mode?

**Comment by Stretchdude** [comment]
I'm in blocking mode. "gatt.connect(address.c_str(), true, false, bleconf->devname);"

My Problem was, that The read_and_process_next was blocking until forever, because the connection wasn't alive. (or maybe some timeout, that feels like forever ;-). I don't know if I just had to have configured an additional callback to solve this. With the boolean stuff I added as workaround It works quite well for me. So I think I can stop bothering you with this.
Maybe this Issue vanishes with the solving of #31?! I'm not sure what leads to the problem in the internals of Libble++. 
I am in summer vacation so I can't do any testing for another week. I'll have a look if there's any dependencies between the two issues soon.

Thanks and regards
Ben
